truffle/com.oracle.truffle.api.instrumentation/src/com/oracle/truffle/api/instrumentation/TruffleInstrument.java

```diff
@@ -0,0 +1,236 @@
+     * exactly once per {@link TruffleInstrument} instance. If the instrument is re-enabled a new
+     * {@link TruffleInstrument} implementation instance is created.
+     *
+     * @param env environment information for the instrumentation
+     */
+    protected void onDispose(Env env) {
+        // default implementation does nothing
+    }
+
+    /**
+     * Represents the environment of a {@link TruffleInstrument}. Provides ways and means to access
+     * input, output and error streams. Also allows to parse arbitrary code from other Truffle
+     * languages.
+     */
+    public static final class Env {
+
+        private final InputStream in;
+        private final OutputStream err;
+        private final OutputStream out;
+
+        Env(OutputStream out, OutputStream err, InputStream in) {
+            this.in = in;
+            this.err = err;
+            this.out = out;
+        }
+
+        /**
+         * Input associated with {@link com.oracle.truffle.api.vm.PolyglotEngine} this
+         * {@link TruffleInstrument instrument} is being executed in.
+         *
+         * @return reader, never <code>null</code>
+         */
+        public InputStream in() {
+            return in;
+        }
+
+        /**
+         * Standard output writer for {@link com.oracle.truffle.api.vm.PolyglotEngine} this
+         * {@link TruffleInstrument instrument} is being executed in.
+         *
+         * @return writer, never <code>null</code>
+         */
+        public OutputStream out() {
+            return out;
+        }
+
+        /**
+         * Standard error writer for {@link com.oracle.truffle.api.vm.PolyglotEngine} this
+         * {@link TruffleInstrument instrument} is being executed in.
+         *
+         * @return writer, never <code>null</code>
+         */
+        public OutputStream err() {
+            return err;
+        }
+
+        /**
+         * Evaluates source of (potentially different) language using the current context.The names
+         * of arguments are parameters for the resulting {#link CallTarget} that allow the
+         * <code>source</code> to reference the actual parameters passed to
+         * {@link CallTarget#call(java.lang.Object...)}.
+         *
+         * @param source the source to evaluate
+         * @param argumentNames the names of {@link CallTarget#call(java.lang.Object...)} arguments
+         *            that can be referenced from the source
+         * @return the call target representing the parsed result
+         * @throws IOException if the parsing or evaluation fails for some reason
+         */
+        @SuppressWarnings("static-method")
+        public CallTarget parse(Source source, String... argumentNames) throws IOException {
+            return InstrumentationHandler.ACCESSOR.parse(null, source, null, argumentNames);
+        }
+
+    }
+
+    /**
+     * Annotation to register {@link TruffleInstrument instrumentation} instances for automatic
```

Comment:
"instrumentation" -> "truffleInstrument" or "instrumentImpl" or "instrument" (or some other variable naming convention I hope we devise - see earlier comment).